### ISSUE TITLE
Support for versions released post 20.04

### ISSUE THREAD

**Issue by JaydenDev** [hakuna-m/wubiuefi#305]
These include
20.10
21.04
21.10
22.04 (Beta)

**Comment by JaydenDev** [comment]
As the downloading of ISOs automatically through the application fails, with an error reported often. ISOs must be downloaded by the user and placed in the directory of the binary win32 executable. However, as the latest supported release is 20.04.4 (the oldest for download), and the ISO looks for 20.04.3, the installation fails.

**Comment by hakuna-m** [comment]
> 20.10
> 21.04

20.10 and 21.04 have reached their end of life. see [Ubuntu releases](https://wiki.ubuntu.com/Releases). So meanwhile, it makes no sense to support an installer of an unsupporfed release.  

> 21.10

As 21.10 is only an interim release with a soon need for an upgrade, we want to push LTS versions instead. Nethertheless, advanced users can use support as described [here](https://github.com/hakuna-m/wubiuefi/wiki#various-distrosversions)  

> 22.04 (Beta)

The download links and the check sums of pre releases are frequently under construction. So our releases in the past for pre releases had frequently download issues. Therefore our decision is to support pre releases without automatic download as described [here](https://github.com/hakuna-m/wubiuefi/wiki#various-distrosversions)

> However, as the latest supported release is 20.04.4 (the oldest for download), and the ISO looks for 20.04.3, the installation fails.

If you disconnect internet connection during first part of installation or if you use `--skipmd5check`as argument ( see [parameters](https://github.com/hakuna-m/wubiuefi/wiki/FAQ#how-can-i-run-wubiuefi-with-parameters-) ) , it should be no problem. Alternately, you can use 2004r344 (https://github.com/hakuna-m/wubiuefi/releases/tag/2004r344) which has no valid isolist for downloads.

Nevertheless, it makes sense to add download support for 20.04.4 and the final releases of 22.04 and we will release it soon.

**Comment by hakuna-m** [comment]
[20044r344](https://github.com/hakuna-m/wubiuefi/releases/tag/20044r344) has been released.

**Comment by Dex94** [comment]
I see actually that  support for ubuntu 22 is removed. Some packages are removed because they considered wubi an obsolete project and they don't know it's still alive.

**Comment by hakuna-m** [comment]
> I see actually that support for ubuntu 22 is removed.

Don't forget, the project Wubiuefi has always existed because of an unsupported and unmaintained Wubi. So it is not a huge surprise that the Ubuntu team does not support Wubi for Ubuntu 22.04.
But of course, it is necessary to review changes which relate Wubiuefi.

> Some packages are removed

The removing of packages is not always an issue for Wubiuefi. Wubiuefi already contains some packages like lupin-support for ISOs without those packages.

**Comment by Dex94** [comment]
I've already tried to install ubuntu 22 and there's an error finding iso path. Someone modified loopback file and said was able to boot the iso. I've tried but nothing.

**Comment by Dex94** [comment]
Today some packages are updated. I'll try to test it

**Comment by Dex94** [comment]
I tried but maybe installation ignores wubi settings. Can you try it?

**Comment by Dex94** [comment]
It ignores wubi settings.

**Comment by Dex94** [comment]
patch/autopartition-loop line 401,  patch/grub-mkimage-lupin line 22, patch/grub-install line 10-12, patch/lupin-grub-install line 10-12, patch/grub-install-efi line 49,114; hooks/post-installer.sh line 19,20,22,24,25,30,32 refers to lupin/lupin support that is a remote memory in ubuntu 22. They need to be modified. Later I'll post a link that I found for a working iso to test. Official one now doesn't work.

**Comment by hakuna-m** [comment]
> Maybe an wubi update could resolve the issue.

Thank you for your reports. As 22.04 has been released since Thursday, we debug the final release of Ubuntu 22.04 at this weekend. I am optimistic that we will find a solution.

**Comment by Dex94** [comment]
happy to help. The problem is now to replace lupin-casper.  If I have seen well, mounting iso instruction is now moved to usr/share/initramfs-tools/scripts/casper-helpers and /usr/share/initramfs-tools/scripts/casper-premount/20iso_scan refers to that file. In ubuntu 20 there is a refer to lupin-helpers, in ubuntu 22 to casper-helpers.

**Comment by dieter-erich** [comment]
Please forgive my impacience but as exited wubi user almost from the beginning (I believe my first version was 10.04) I am eager to install wubi for mate 22.04. Is there a projected date for its release? Thanks to all who contributed to this often reviled but fantastic system!

**Comment by hakuna-m** [comment]
Sorry, for my late reply. I have been sick for a week. 

An upgrade to 22.04 seems to work. see https://github.com/hakuna-m/wubiuefi/issues/308#issuecomment-1107367059

Currently, we test Ubuntu 22.04 where bug [1960457](https://bugs.launchpad.net/ubuntu/+source/lupin/+bug/1960457) has been fixed.

If it works we will test other flavours like Ubuntu Mate 22.04.

**Comment by hakuna-m** [comment]
The Ubuntu team has removed boot option "debian-installer/custom-installation=" in Ubuntu 22.04. It seems that the team does not want that someone can easily customize the ISO.

**Comment by Dex94** [comment]
Can be booted custom iso?

**Comment by dieter-erich** [comment]
Thanks again HM! I again tried installing mate 20.04 (to be finally 
release-upgraded to 22.04) on my HP Z book: Wubi runs OK but on re-boot 
I invariably arrive at "> grub". I then tried to disable secure boot but 
whatever I do I am sent to a screen telling me that I have to enter the 
recovery key etc. The only way to save my Win10 install was to 
re-activate secure boot in the boot menu. Any ideas? It appears to me 
that it becomes more and more difficult to change anything in the 
Windows system. There is also a lot of security stuff coming from HP. Is 
there any EASY way to make wubi running on this Z book? In the next days 
I'll try release-upgrading on other laptops as well.

------------------------------------------------------------------------
Dieter Blaas,
Max Perutz Laboratories
Medical University of Vienna,
Inst. Med. Biochem., Vienna Biocenter (VBC),
Dr. Bohr Gasse 9/3,
A-1030 Vienna, Austria,
Tel: 0043 1 4277 61630,
Mobile: 0043 699 1942 1659
***@***.***
------------------------------------------------------------------------

**Comment by Dex94** [comment]
I see  that in some hdd is difficult to boot ubuntu from windows manager and it's needed to put  the priority to grub. Try to disable secure boot and enter directly to grub pressing on boot the relative button to choose the boot manager or choosing it directly from bios. There's written something like the name of the distro.

**Comment by hakuna-m** [comment]
> Can be booted custom iso?

An ISO with modified initrd file still works.  

But good news. We have found a solution to substitute missing boot option "debian-installer/custom-installation=" with other existing boot options.

**Comment by hakuna-m** [comment]
> on my HP Z book

Your HP Z book seems to have special issues as you wrote in #303 . First of all, we try to fix general issues for Ubuntu 22.04.

**Comment by Dex94** [comment]
> > Can be booted custom iso?
> 
> An ISO with modified initrd file still works.
> 
> But good news. We have found a solution to substitute missing boot option "debian-installer/custom-installation=" with other existing boot options.

can you summarize what the problem was in casper and in the boot option? for curiosity...

**Comment by hakuna-m** [comment]
To put it in a nutshell, the boot option "debian-installer/custom-installation=" provides the content of the path (parameter value in our case is **/ubuntu/install/custom-installation**) in **/custom-installation**.

If there is no **/custom-installation** during boot time all scripts containing path name beginning with /custom-installation fail.

The first script which fails is that one in preseed.cfg
```
## SHELL COMMANDS.
d-i preseed/early_command string [ -x /custom-installation/hooks/early-command.sh ] && /custom-installation/hooks/early-command.sh
```
So it is necessary to rewrite that part.

**Comment by ptog** [comment]
hello
the 20.04 kernel in the latest wubi release is not recognizing the wifi adapter (intel killer wifi 6E AX1675) on my msi notebook, hence networking doesn't work at all which also prevents the mentioned "sudo do-release-upgrade -d" solution...
support for a later ubuntu like 22.04 (hopefully?) would be great
i really quite love the wubi approach! (happily using it with 20.04 on a lenovo thinkbook). many thanks for your work : )

**Comment by JaydenDev** [comment]
![image](https://user-images.githubusercontent.com/92550746/168906232-bedcbe80-3723-4d77-ad4b-253f461613e8.png)
[Source](https://www.intel.com/content/www/us/en/support/articles/000088040/wireless.html)

The generic Linux drivers should work.

**Comment by hakuna-m** [comment]
An update with the status report:

- [x] change ISO downloads to 22.04 (data/isolist.ini)
- [x] change boot option "debian-installer/custom-installation=" to "file=" (data/grub.install.cfg)
- [x] change preseed file (data/preseed.lupin)
- [x] add missing features to hooks/early-command.sh
- [x] test scripts within initramfs environment of 22.04 (works)
- [x] build new version
- [x] test Windows part (works)
- [ ] test Linux part in a virtual machine
- [ ] release new version 

So probably tomorrow, we will have a tested new version.

BTW, in our tests we noted that boot option "iso-scan/filename=" has also changed. In 22.04 the mount point **/isodevice** is not available in live environment of the ISO. It is also tricky to use it with boot option "file=".

**Comment by hakuna-m** [comment]
Here are the results of the tests in a VM:

The fix for the changed boot options works, The installer uses our configuration.
But we need a second fix for the missing package lupin-support. The existing fix does not work. But it should be no huge problem.

**Comment by hakuna-m** [comment]
New decision: As package lupin-support has been removed we rewrite the Wubi patches. The new patches will not fix issues in lupin-support, they will fix issues directly in Grub.

This has the huge advantage that we need no package lupin-support and we can better avoid Grub issues after upgrades or installs.

**Comment by hakuna-m** [comment]
Good news. It works. So we can release a version for Ubuntu 22.04.

- [x] test Linux part in a virtual machine (works)
- [ ] release new version (today)

**Comment by JaydenDev** [comment]
> Good news. It works. So we can release a version for Ubuntu 22.04.
> 
>     * [x]  test Linux part in a virtual machine (works)
> 
>     * [ ]  release new version (today)

Excellent! I'll be happy to test it directly after release.

**Comment by golgeadam** [comment]
hello, is there a linux mint md5 compatible wubi version for win11. success in your studies

**Comment by hakuna-m** [comment]
> I again tried installing mate 20.04 (to be finally release-upgraded to 22.04) on my HP Z book: Wubi runs OK but on re-boot I invariably arrive at "> grub". 

see https://github.com/hakuna-m/wubiuefi/issues/308#issuecomment-1133958135

> hello, is there a linux mint md5 compatible wubi version for win11. success in your studies

see https://github.com/hakuna-m/wubiuefi/issues/310#issuecomment-1133952582

**Comment by dieter-erich** [comment]
Hi,

      I now tried installing Ubuntu-mate 22.04 with the brand new 
wubi-22.04 on my HP Z-book running win10: After finishing the 
installation and rebooting I was again sent to the grub prompt. Again 
rebooting, Windows started normally but did not show any "ubuntu" and no 
boot menu. In C:\ there is a directory "ubuntu" with everything in it 
and there are the files wubildr and wubildr.mdr. Could it be that the 
wubi files have to go into the EFI partition (see screenshot)? As far as 
I understand WUBI does not find the files it needs to boot from. But, is 
it allowable to copy them into the EFI partition? I do not want to mess 
up my Windows!

------------------------------------------------------------------------
Dieter Blaas,
Max Perutz Laboratories
Medical University of Vienna,
Inst. Med. Biochem., Vienna Biocenter (VBC),
Dr. Bohr Gasse 9/3,
A-1030 Vienna, Austria,
Tel: 0043 1 4277 61630,
Mobile: 0043 699 1942 1659
***@***.***
------------------------------------------------------------------------

**Comment by Dex94** [comment]
in efi partition sometimes there are some problems. In a computer I need to press a key, I don't remenber if it's esc or f10 to select grub boot otherwise it doesn't see ubuntu. Try to boot from grub.

**Comment by dieter-erich** [comment]
When I go into the boot menu by pressing esc several times I see, among 
other options, "UEFI Wubi". When I press it, I am again sent to >grub

------------------------------------------------------------------------
Dieter Blaas,
Max Perutz Laboratories
Medical University of Vienna,
Inst. Med. Biochem., Vienna Biocenter (VBC),
Dr. Bohr Gasse 9/3,
A-1030 Vienna, Austria,
Tel: 0043 1 4277 61630,
Mobile: 0043 699 1942 1659
***@***.***
------------------------------------------------------------------------

**Comment by Dex94** [comment]
is secureboot disabled?

**Comment by Dex94** [comment]
> Good news. It works. So we can release a version for Ubuntu 22.04.
> 
>     * [x]  test Linux part in a virtual machine (works)
> 
>     * [x]  release new version (today)

have you tried only ubiquity or other installer?

**Comment by dieter-erich** [comment]
No, because if I do this it screws up my Windows start - gives a lot of 
error messages and finally it boots Windows. This only works when I 
allow system restore (or something similar). I am too afraid to 
permanently damage Windows start....

------------------------------------------------------------------------
Dieter Blaas,
Max Perutz Laboratories
Medical University of Vienna,
Inst. Med. Biochem., Vienna Biocenter (VBC),
Dr. Bohr Gasse 9/3,
A-1030 Vienna, Austria,
Tel: 0043 1 4277 61630,
Mobile: 0043 699 1942 1659
***@***.***
------------------------------------------------------------------------

**Comment by hakuna-m** [comment]
> have you tried only ubiquity or other installer?

Only Ubuntu 22.04 with Ubiquity. Other installers are a konwn issue which should work for e.g. Lubuntu with rev. 346.

**Comment by hakuna-m** [comment]
> As far as I understand WUBI does not find the files it needs to boot from.

@dieter-erich I edited my comment for GRUB issues https://github.com/hakuna-m/wubiuefi/issues/308#issuecomment-1133958135

It is not necessary to copy files to EFI partition after installation. 

The boot sequence for a working Wubiuefi with Ubuntu 22.04:

UEFI boot menu -> /efi/ubuntu/wubildr/shimx64.efi from EFI partition -> /efi/ubuntu/wubildr/grubx64.efi from EFI partition -> /boot/grub/grub.cfg from /ubuntu/disks/root.disk from Windows drive C: -> /boot/vmlinuz-5.15.0-25-generic + /boot/initrd.img-5.15.0-25-generic from /ubuntu/disks/root.disk from Windows drive C:
